gr-fec/lib/tagged_decoder_impl.cc

```diff
@@ -51,7 +51,9 @@ int tagged_decoder_impl::calculate_output_stream_length(const gr_vector_int& nin
     if ((ninput_items[0] * d_decoder->rate()) > (d_mtu * 8)) {
         throw std::runtime_error("tagged_encoder: received frame is larger than MTU.");
     }
-    d_decoder->set_frame_size(round(ninput_items[0] * d_decoder->rate()));
+    int diff =
+        d_decoder->rate() * d_decoder->get_input_size() - d_decoder->get_output_size();
+    d_decoder->set_frame_size(round(ninput_items[0] * d_decoder->rate()) - diff);
```

Comment:
This is a change in the wrapper. Not the CC decoder itself. Other decoders might return `false` in case the `set_frame_size` method is called.

Maybe the decoder should compute this kind of info internally since it is only required in the CC case but LDPC, polar etc. don't need it.

Or, we need a second method that sets the input size rather than the output size. For the CC case that would probably make more sense. output size == frame size.

Reply:
This kind of code also appears on `async_decoder`, which is also the wrapper, not the CC decoder:

https://github.com/gnuradio/gnuradio/blob/7f1a32432fd45924a9eb5f3e3cc99d3248b740f7/gr-fec/lib/async_decoder_impl.cc#L86-L93

So I think that whatever solution we deem appropriate here (such as adding a method for the input size) would also need to be applied to the `async_decoder`.

Reviewing the code of the different wrappers and the implementation of the CC decoder, I agree that there are several parts that would improve from a refactor.

Since currently `tagged_decoder` is broken and I think that `async_decoder` is not broken, I think there are two possible ways to go:
1. Merge this PR, keep the idea to improve the code quality in the future.
2. Don't merge this PR. Wait until a refactor/improvement of the code quality that also fixes this problem as a by-product.

I prefer approach 1, because it would fix the problem described in this PR right away and it doesn't get in the way of other refactors/improvements happening afterwards. With approach 2, it might be a long time until this particular problem gets fixed, because the best way to improve the code is not clear at the moment.

The only potential concern that I have about merging this PR is whether it breaks something else. I think it doesn't, but I haven't reviewed the other decoders to make sure that this change is harmless for them.

Reply:
The second approach is effectively an API change. This would require a new release. I'd prefer the second approach but I agree that your approach is the better one for now.

The FECAPI needs some refactoring. i.e. a better understanding of codes and properties. Currently, the abstraction is broken.